### DISCUSSION TITLE
es/modes/track/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#8474] [Docs]
# es/modes/track/

Aprende a utilizar Ultralytics YOLO  para el seguimiento de objetos en secuencias de vídeo. Guías para utilizar diferentes rastreadores y personalizar las configuraciones de los rastreadores.

https://docs.ultralytics.com/es/modes/track/

**Comment by wisrovi** [comment]
Hello

Is possible to get the result in dataframe?

for example:
I have a video and applied track, and i need to extract the filter the track by each class, my model looking for car and persons

I like to get the result for to filter for ids for every class

**Reply by pderrenger** [reply]
@wisrovi ¡Hola! 😊 Sí, es posible obtener los resultados en un DataFrame de pandas para filtrar por clase y ID. Aquí tienes un ejemplo de cómo hacerlo después de ejecutar el seguimiento en un video:

```python
import pandas as pd
from ultralytics import YOLO

# Cargar el modelo y realizar el seguimiento
model = YOLO('yolov8n.pt')  # Asegúrate de usar tu modelo correcto aquí
results = model.track(source='path/to/your/video.mp4')

# Crear un DataFrame vacío
df = pd.DataFrame(columns=['frame', 'track_id', 'class', 'x', 'y', 'width', 'height'])

# Llenar el DataFrame con resultados
for i, result in enumerate(results):
    for j, (box, cls, track_id) in enumerate(zip(result.boxes.xyxy, result.boxes.cls, result.boxes.id)):
        df = df.append({'frame': i, 'track_id': track_id.item(), 'class': result.names[cls.item()], 
                        'x': box[0].item(), 'y': box[1].item(), 'width': box[2].item() - box[0].item(), 
                        'height': box[3].item() - box[1].item()}, ignore_index=True)

# Filtrar por clase, por ejemplo, 'car' y 'person'
df_car = df[df['class'] == 'car']
df_person = df[df['class'] == 'person']

# Ahora df_car y df_person contienen los rastreos filtrados por clase
print(df_car.head())
print(df_person.head())
```

Este script carga tu modelo, realiza el seguimiento en el video especificado y luego llena un DataFrame con los resultados, incluyendo el ID de seguimiento, la clase y las coordenadas del cuadro delimitador. Luego, puedes filtrar fácilmente este DataFrame por la clase que te interese. Espero que esto te ayude. ¡Déjanos saber si tienes más preguntas! 🚀

**Reply by wisrovi** [reply]
thanks

**Reply by pderrenger** [reply]
@wisrovi ¡Hola! 👋 Para filtrar los resultados de seguimiento por clase y obtener los IDs para cada clase (por ejemplo, 'car' y 'person'), puedes hacerlo directamente después de ejecutar el seguimiento. Aquí te dejo un ejemplo conciso usando pandas para organizar y filtrar los resultados:

```python
import pandas as pd
from ultralytics import YOLO

# Cargar el modelo y realizar el seguimiento en tu video
model = YOLO('yolov8n.pt')  # Asegúrate de usar tu modelo correcto aquí
results = model.track(source='path/to/your/video.mp4')

# Crear un DataFrame para almacenar los resultados
tracks_df = pd.DataFrame([{
    'frame_index': i,
    'track_id': box.id.item(),
    'class': result.names[box.cls.item()],
    'x': box.xyxy[0].item(),
    'y': box.xyxy[1].item(),
    'width': box.xyxy[2].item() - box.xyxy[0].item(),
    'height': box.xyxy[3].item() - box.xyxy[1].item()
} for i, result in enumerate(results) for box in result.boxes])

# Filtrar por clase
cars_df = tracks_df[tracks_df['class'] == 'car']
persons_df = tracks_df[tracks_df['class'] == 'person']

# Ahora tienes dos DataFrames: cars_df y persons_df, que puedes usar para análisis adicional
print(cars_df.head())
print(persons_df.head())
```

Este script te permitirá tener un DataFrame con todos los tracks y luego filtrarlos por las clases que te interesan. ¡Espero que esto te sea útil! 😊

**Comment by wisrovi** [comment]
No, yes, I can post this question here.

I have a segmentation model that I am using to track each instance, I am trying to convert each mask into a shapely polygon (from shapely.geometry import Polygon)

but I'm not getting it, for some reason, I can't get the conversion, is there a way to convert it directly or how would you recommend doing the conversion?

for example, I have:

results = model.track(source=im0)

I tried with:

all_masks = []
            orig_shape = result.orig_shape[::-1]
            for mask in result.masks:
                if mask is not None:
                    p = mask.data.cpu().numpy()[0]
                    p = cv2.resize(p, orig_shape)
                    
                    contours = cv2.findContours(p.astype(np.uint8),cv2.RETR_EXTERNAL,cv2.CHAIN_APPROX_SIMPLE,)[0]

but, It is not working

**Reply by pderrenger** [reply]
¡Hola! Para convertir las máscaras de segmentación en polígonos usando Shapely, primero necesitas asegurarte de que estás extrayendo correctamente los contornos de las máscaras con OpenCV. Luego, puedes convertir estos contornos en polígonos de Shapely. Aquí te dejo un ejemplo de cómo podrías hacerlo:

```python
from shapely.geometry import Polygon
import cv2
import numpy as np

# Suponiendo que 'p' es tu máscara binaria después de redimensionar
# y que ya tienes 'contours' como en tu código

for contour in contours:
    # Convertir contorno en polígono de Shapely
    # Asegúrate de que 'contour' es la forma correcta (n, 1, 2)
    contour = contour.squeeze()
    polygon = Polygon(contour)
    print(polygon)
```

Si `cv2.findContours` no está funcionando como esperas, verifica que la máscara 'p' sea binaria (valores 0 y 255) y que estés utilizando correctamente los parámetros de `cv2.findContours`. También, asegúrate de que `contour.squeeze()` esté dando la forma correcta para crear el polígono en Shapely.

Si sigues teniendo problemas, podría ser útil revisar el tipo de datos y la forma de la máscara 'p' después de redimensionarla y antes de aplicar `findContours`. A veces, pequeños detalles en estos pasos pueden afectar el resultado final.

Espero que esto te ayude a avanzar con tu proyecto. ¡Mucho éxito! 🚀
